File: appkit/src/main/scala/org/ergoplatform/appkit/ColdErgoClient.scala
package org.ergoplatform.appkit

import java.util.function
import org.ergoplatform.restapi.client
import org.ergoplatform.appkit.impl.{ColdBlockchainContext, NodeInfoParameters}
import org.ergoplatform.restapi.client.NodeInfo

class ColdErgoClient(networkType: NetworkType, params: BlockchainParameters) extends ErgoClient {

  /**
    * Convenience constructor for giving maxBlockCost
    */
  def this(networkType: NetworkType, maxBlockCost: Int, blockVersion: Byte) {
    this(networkType, new NodeInfoParameters(
      new NodeInfo().parameters(new client.Parameters()
        .maxBlockCost(Integer.valueOf(maxBlockCost))
        .blockVersion(Integer.valueOf(blockVersion)))))
  }

  override def execute[T](action: function.Function[BlockchainContext, T]): T = {
    val ctx = new ColdBlockchainContext(networkType, params)
    val res = action.apply(ctx)
    res
  }
  override def getDataSource: BlockchainDataSource = ???
}

File: appkit/src/main/java/org/ergoplatform/appkit/ColdErgoClient.java
package org.ergoplatform.appkit;

import org.ergoplatform.appkit.impl.ColdBlockchainContext;
import org.ergoplatform.appkit.impl.NodeInfoParameters;
import org.ergoplatform.restapi.client.NodeInfo;
import org.ergoplatform.restapi.client.Parameters;

import java.util.function.Function;

public class ColdErgoClient implements ErgoClient {

    private final NetworkType networkType;
    private final BlockchainParameters params;

    public ColdErgoClient(NetworkType networkType, BlockchainParameters params) {
        this.networkType = networkType;
        this.params = params;
    }

    /**
     * Convenience constructor for giving maxBlockCost
     */
    public ColdErgoClient(NetworkType networkType, int maxBlockCost, byte blockVersion) {
        this(networkType, new NodeInfoParameters(
            new NodeInfo()
                .parameters(new Parameters()
                    .maxBlockCost(maxBlockCost)
                    .blockVersion((int) blockVersion))
        ));
    }

    public BlockchainParameters params() {
        return params;
    }

    @Override
    public <T> T execute(Function<BlockchainContext, T> action) {
        ColdBlockchainContext ctx = new ColdBlockchainContext(networkType, params);
        return action.apply(ctx);
    }

    @Override
    public BlockchainDataSource getDataSource() {
        throw new UnsupportedOperationException();
    }
}
